src/NHibernate/Linq/Visitors/QueryModelVisitor.cs

```diff
@@ -551,6 +556,54 @@ private void AddJoin(JoinClause joinClause, QueryModel queryModel, bool innerJoi
 			_hqlTree.AddFromClause(join);
 		}
 
+		private class BaseMemberChecker : NhExpressionVisitor
+		{
+			private readonly ISessionFactoryImplementor _sessionFactory;
+			private bool _result;
+
+			public BaseMemberChecker(ISessionFactoryImplementor sessionFactory)
+			{
+				_sessionFactory = sessionFactory;
+			}
+
+			public bool ContainsBaseMember(JoinClause joinClause)
+			{
+				// Visit the join inner key only for entities that have subclasses
+				if (joinClause.InnerSequence is ConstantExpression constantNode &&
```

Comment:
Constant check seems to be enough as `join o2 in db.Mammals.Where(o => o.Pregnant)` is not supported.